src/libopensc/card-cac.c

```diff
@@ -1008,6 +1096,130 @@ static int cac_decipher(sc_card_t *card,
 	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, cac_rsa_op(card, data, datalen, out, outlen));
 }
 
+static int parse_properties_object(sc_card_t *card, u8 type,
+    u8 *data, size_t data_len, cac_properties_object_t *object)
+{
+	size_t len;
+	u8 *val, *val_end, tag;
+	int parsed = 0;
+
+	if (data_len < 11)
+		return -1;
+
+	/* Initilize: non-PKI applet */
+	object->privatekey = 0;
+
+	val = data;
+	val_end = data + data_len;
+	for (; val < val_end; val += len) {
+		/* get the tag and the length */
+		if (sc_simpletlv_read_tag(&val, val_end - val, &tag, &len) != SC_SUCCESS)
+			break;
+
+		switch (tag) {
+		case CAC_TAG_OBJECT_ID:
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: Object ID = 0x%02x 0x%02x", val[0], val[1]);
+			memcpy(&object->oid, val, 2);
+			parsed++;
+			break;
+
+		case CAC_TAG_BUFFER_PROPERTIES:
+			/* First byte is "Type of Tag Supported" */
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: Buffer Properties: Type of Tag Supported = 0x%02x",
+			    val[0]);
+			object->simpletlv = val[0];
+			parsed++;
+			break;
+
+		case CAC_TAG_PKI_PROPERTIES:
+			/* 4th byte is "Private Key Initialized" */
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: PKI Properties: Private Key Initialized = 0x%02x",
+			    val[2]);
+			object->privatekey = val[2];
```

Comment:
This is still not checking the length, right? Here, you should check whether `len >= 3`. For the other cases above it's similar...

Reply:
True. The specification clearly says that this tag should have 4 bytes so explicit check for 4B length makes sense here. Thanks.